### ISSUE TITLE
Standardize discovery, installation and usage of LLM functions, agents and tools

### ISSUE THREAD

**Issue by pierDipi** [kserve/kserve#3829]
/kind feature

**Describe the solution you'd like**

This proposal outlines the need for an API to standardize the discovery, installation, and aggregation of LLM functions, agents or tools in Kubernetes.

LLM agents are a powerful extension to RAG-based LLM applications. They enable complex tasks that require calculations, interaction with external systems, and actions beyond information retrieval.

## Example

- User Prompt:

> How has the trend in the average resource consumption (CPU, GPU, memory) changed over the past 2 months for my application, and what impact might this have on my team cloud bill?

Analyzing application resource consumption trends and cost implications requires calculations and graphical representation, tasks not well-suited for RAG. LLM agents can address this by:

- Employing a monitoring agent to write and execute queries for resource usage data.
- Utilizing a cost management agent to estimate costs based on usage data.

This approach minimizes the risk of inaccurate responses and allows users to ask various questions about resource usage and costs.

Utilizing agents currently requires manual and bespoke API wrapping for various deployments and serverless services or using proprietary APIs. This hinders user experience and limits adoption.

![Agentic LLM in OCP and_or Podman](https://github.com/user-attachments/assets/76bf8a01-5fd3-4996-a579-8ec216155a83)

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

We've created a POC based on the Knative Eventing EventType API here https://github.com/keventmesh/llm-tool-provider, however, We're looking for a more independent project as the solution is generally applicable beyond Knative.

The POC allows to dynamically attach agents and external services to a chat application and is able to resolve the above example user prompt.

You can read more about this in these 2 blog posts:
- https://knative.dev/blog/articles/llm-agents-overview/
- https://knative.dev/blog/articles/llm-agents-demo/

**Links to the design documents:**

- (early draft) https://docs.google.com/document/d/1ihDaVcs3TfhQEU92Xkzh5vq6PQoxUS81F81WV4Vnp_I/edit

**Comment by pierDipi** [comment]
See @cali0707 demo of the POC for CNCF AI WG https://youtu.be/-sWl4XBRsFs?si=FvQhLfJB-B-ojs9B&t=3363

**Comment by lizzzcai** [comment]
One feature I can think of for tool calling is that the inference service deployed via KServe (KServe SDK) should have a schema endpoint (example: `/schema`) that returns its schema and description (auto-generated) so the schema and description are not needed [here](https://knative.dev/blog/articles/llm-agents-demo/#:~:text=knative.dev/v1beta2%0Akind%3A-,EventType,-metadata%3A%0A%20%20name%3A%20average). 
Agent service just need to know the inference service/endpoint and retrieve the function spec and description from the endpoint itself.

**Comment by yuzisun** [comment]
@pierDipi This does look like in the scope of KServe and we have roadmap to support RAG/Agent flow by extending KServe inference graph API. Would you like to discuss this further in the KServe community meeting?

**Comment by pierDipi** [comment]
Yes, Dan, I can start the discussion in one of the next community meeting

**Comment by terrytangyuan** [comment]
vLLM added support for [OpenAI-Compatible Tools API](https://github.com/vllm-project/vllm/pull/5649)

**Comment by andyi2it** [comment]
@pierDipi 
Are you actively working on this?

**Comment by pierDipi** [comment]
@andyi2it last time we met, we wanted to gather more feedback and whether it's a project that is a good fit for KServe, I believe it is and helps with opening up new use cases for Kserve but I'm not sure if KServe leads have decided this is a project that they would host.

If that's the case, I would love to start working on it.

@yuzisun in the meeting, we were discussing that you wanted to discuss agentic use cases internally, did you get any feedback?

**Comment by pierDipi** [comment]
@andyi2it would you mind sharing more your use cases? It can help with shaping the project better

**Comment by pierDipi** [comment]
@cali0707 presented a lighting talk at KubeCon featuring the prototype here: https://www.youtube.com/watch?v=wHxIaZawes8
